### PR TITLE
new feature: update ellipse detector using projective invariant pruning

### PR DESCRIPTION
**Pull request by MisakiCoca**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/ximgproc/include/opencv2/ximgproc/ellipse_detector.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+* [JIA2017FAST]  Jia, Qi et al, (2017). A Fast Ellipse Detector using Projective Invariant Pruning. IEEE Transactions on Image Processing.
+*
+@param _image input image, could be gray or color.
+@param _ellipses output vector of found ellipses. each vector is encoded as five float $x, y, a, b, radius, score$.
+@param scoreThreshold float, the threshold of ellipse score.
+@param reliabilityThreshold float, the threshold of reliability.
+@param centerDistanceThreshold float, the threshold of center distance.
+*/
+CV_EXPORTS_W void ellipseDetector(
+    InputArray _image, OutputArray _ellipses,
```
Comment (sturkmen72) [inline modules/ximgproc/include/opencv2/ximgproc/ellipse_detector.hpp:30]: Please change like
InputArray image, OutputArray ellipses

File: modules/ximgproc/include/opencv2/ximgproc/ellipse_detector.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+@brief finds ellipses fastly in an image using projective invariant pruning.
+*
+* The function detects ellipses in images using projective invariant pruning.
+* For more details about this implementation, please see
+* [JIA2017FAST]  Jia, Qi et al, (2017). A Fast Ellipse Detector using Projective Invariant Pruning. IEEE Transactions on Image Processing.
+*
+@param _image input image, could be gray or color.
+@param _ellipses output vector of found ellipses. each vector is encoded as five float $x, y, a, b, radius, score$.
```
Comment (sturkmen72) [inline modules/ximgproc/include/opencv2/ximgproc/ellipse_detector.hpp:24]: Please change like
@param image ...
@param ellipses ...

File: modules/ximgproc/src/ellipse_detector.cpp [hunk 3]
```diff
@@ -0,0 +1,2015 @@
...
+
+        // create a new cluster
+        if (!foundCluster)
+            clusters.push_back(e1);
+    }
+    clusters.swap(ellipses);
+}
+
+// enter of ellipse detector
+CV_EXPORTS void ellipseDetector(
```
Comment (sturkmen72) [inline modules/ximgproc/src/ellipse_detector.cpp:1972]: No need to CV_EXPORTS here

**Comment by MisakiCoca** [comment]
@sturkmen72 Hello, we have passed the test and the code has been modified according to your instructions.

**Comment by sturkmen72** [comment]
@MisakiCoca i am a common contributor. a developers team member will review.

consider my additional remarks as follows :

1. ellipseDetector() seems a class name. maybe something like `findEllipses` will be better.
2. you can see the preview of the documentation https://pullrequest.opencv.org/buildbot/export/pr_contrib/3322/docs/d1/d05/group__ximgproc__ellipse__detector.html#ga4f98330bb312dd8f42046105e646b390

just for a function I think no need to add a section in modules part https://pullrequest.opencv.org/buildbot/export/pr_contrib/3322/docs/df/d2d/group__ximgproc.html

**Comment by MisakiCoca** [comment]
> @MisakiCoca i am a common contributor. a developers team member will review.
> 
> consider my additional remarks as follows :
> 
> 1. ellipseDetector() seems a class name. maybe something like `findEllipses` will be better.
> 2. you can see the preview of the documentation https://pullrequest.opencv.org/buildbot/export/pr_contrib/3322/docs/d1/d05/group__ximgproc__ellipse__detector.html#ga4f98330bb312dd8f42046105e646b390
> 
> just for a function I think no need to add a section in modules part https://pullrequest.opencv.org/buildbot/export/pr_contrib/3322/docs/df/d2d/group__ximgproc.html
> 
> 3. I did not test your code but what is your opinion about
> 
> yours : `EllipseDetectorTest.ManySmallEllipses (101787 ms)` and existing ellipse finding EdgeDrawing : `test ximgproc_ED.ManySmallCircles (875 ms)`

Thank you very much for your reply.

We have completed the revisions for the first and second points based on your advice.

Regarding the third point about our running speed: 

Our approach can be roughly summarized as enumerating the arcs in each quadrant and applying a pruning strategy. 
Thus when there are too many arcs, our speed will be slower than algorithms such as Hough, who are based on point sampling. In such situation, however, our method will have higher accuracy. 

In thress realistic datasets with moderate amount of ellipses, the advantages of both the computational speed and the accuracy of our method are very obvious, as shown in the following Figure.
Figure.1
Where the`RHT` and `Qi et al.,` denotes the Random Hough Transform and ours, respectively.

Compared to some other arc-based approach, our computational speed has been improved significantly while retaining accuracy.

Figure.2
Where the `Jia` denotes our method. Pre, Rec are abbreviations of precision and recall.

We have removed the `ManySmallEllipses` from tests, we would be honored if these descriptions could answer your questions.

File: modules/ximgproc/src/find_ellipses.cpp [hunk 4]
```diff
@@ -0,0 +1,2012 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+#include <opencv2/core.hpp>
+#include <unordered_map>
+
```
Comment (sturkmen72) [inline modules/ximgproc/src/find_ellipses.cpp:8]: to compile on windows i added 
`#include <numeric>`

File: modules/ximgproc/samples/find_ellipses.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#include <opencv2/imgproc.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/ximgproc.hpp>
+
+using namespace cv;
+
+int main() {
+
+    // load image
+    Mat img = imread("test.jpg", IMREAD_COLOR);
```
Comment (sturkmen72) [inline modules/ximgproc/samples/find_ellipses.cpp:15]: i suggest
`Mat img = imread(samples::findFile("board.jpg"), IMREAD_COLOR);`

File: modules/ximgproc/samples/find_ellipses.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        ellipse(
+                img,
+                Point(cvRound(ell[0]), cvRound(ell[1])),
+                Size(cvRound(ell[2]), cvRound(ell[3])),
+                ell[5] * 180 / CV_PI, 0.0, 360.0, color, 3
+        );
+    }
+
+    // save image
+    imwrite("result.jpg", img);
```
Comment (sturkmen72) [inline modules/ximgproc/samples/find_ellipses.cpp:44]: i suggest
    imshow("result", img);
    waitKey();

File: modules/ximgproc/src/find_ellipses.cpp (hunk 4 above)
Comment (sturkmen72) [reply modules/ximgproc/src/find_ellipses.cpp:8]: i tried your sample find_ellipses.cpp
here is the result. i wonder if your algorithm finds circles or not. or can find circles ?
result

Comment (MisakiCoca) [reply modules/ximgproc/src/find_ellipses.cpp:8]: Thank you very much for your reply, our method can find out circles in images.

However, for this schematic, it is fatal for all the arc-based ellipse detection algorithms due to the small scale of some of the ellipses and the difficulty of edge detection. The best result we obtained after adjusting the threshold for this sample is shown in the following figure.
Figure.A
Usually, for this case, we recommend to upsample the image twice or more and then detect it, there will be a slight improvement in such situation.

The detection of small ellipses has been a problem in this field, and we have done experiments for this. As can be seen from the figure, such methods will not be able to detect when the length of the long axis of a certain ellipse is less than 25 pixels.
Figure.B

In addition, we have also attached one of the more used images in the field and its detection results to demonstrate that our method works for circles.
Figure.C

We would be honored if these descriptions could answer your questions.

File: modules/ximgproc/samples/find_ellipses.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <iostream>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/imgcodecs.hpp>
```
Comment (sturkmen72) [inline modules/ximgproc/samples/find_ellipses.cpp:7]: #include <opencv2/highgui.hpp> needed

Comment (asmorkalov) [reply modules/ximgproc/samples/find_ellipses.cpp:7]: Highgui dependency for tests should be optional. It's not available on many platforms.

Comment (sturkmen72) [reply modules/ximgproc/samples/find_ellipses.cpp:7]: This is sample, not a test.

**Comment by MisakiCoca** [comment]
@asmorkalov Thank you very much for your attention! What can i do for the macOS-ARM64 failure?

**Comment by asmorkalov** [comment]
It looks like CI issue or system configuration issue on our side. Please ignore it for now.

**Comment by MisakiCoca** [comment]
@asmorkalov Hello, is there anything I can do to aid in the merging of this PR?

File: modules/ximgproc/src/find_ellipses.cpp [hunk 8]
```diff
@@ -0,0 +1,2013 @@
...
+namespace ximgproc {
+
+// typedef
+typedef std::vector<Point> VP;
+typedef std::vector<VP> VVP;
+typedef unsigned int uint;
+
+// ellipse format
+struct Ellipse {
+    float center_x, center_y;
```
[the remark this answers is not in the document]
Comment (MisakiCoca) [reply modules/ximgproc/src/find_ellipses.cpp:20]: Thanks, as you indicated, I will modify these codes as soon as possible.

**Comment by MisakiCoca** [comment]
@asmorkalov Hello, I have completed all the modifications you suggested.

**Comment by MisakiCoca** [comment]
@asmorkalov Sorry, there were some issues with the previous testing session, but they have already been resolved.

**Comment by MisakiCoca** [comment]
@asmorkalov The warning from clang in macOS caused by C.21 have been solved.

**Review by asmorkalov — APPROVED** [review]
Looks good to me in general besides `srand`.

File: modules/ximgproc/src/find_ellipses.cpp [hunk 9]
```diff
@@ -0,0 +1,1990 @@
...
+    _maxCenterDistance = 100.0f * 0.05f;
+    _maxCenterDistance2 = _maxCenterDistance * _maxCenterDistance;
+    _minEdgeLength = 16;
+    _minOrientedRectSide = 3.0f;
+    _maxDistanceToEllipseContour = 0.1f;
+    _minScore = 0.7f;
+    _minReliability = 0.5;
+    _uNs = 16;
+
+    srand(unsigned(time(nullptr)));
```
[the remark this answers is not in the document]
Comment (MisakiCoca) [reply modules/ximgproc/src/find_ellipses.cpp]: Thanks for your reminder! We have removed `srand` from code.

**Comment by MisakiCoca** [comment]
@asmorkalov Thank you so much for your continued help!

File: modules/ximgproc/test/test_find_ellipses.cpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+#include "test_precomp.hpp"
+
+using namespace cv;
+
+namespace opencv_test { namespace {
+
+
+TEST(FindEllipsesTest, EllipsesOnly)
```
Comment (alalek) [inline modules/ximgproc/test/test_find_ellipses.cpp:11]: In debug build mode there is catched out of boundary access issue:

```
[ RUN      ] FindEllipsesTest.EllipsesOnly
unknown file: Failure
C++ exception with description "OpenCV(4.6.0-dev) /build/precommit-contrib_linux64_no_opt/4.x/opencv/modules/core/include/opencv2/core/mat.inl.hpp:1603: error: (-215:Assertion failed) 0 <= y && y < size.p[0] in function 'operator[]'
" thrown in the test body.
[  FAILED  ] FindEllipsesTest.EllipsesOnly (18 ms)
```

@misakicoca Could have a chance to look on this?

Comment (misakicoca) [reply modules/ximgproc/test/test_find_ellipses.cpp:11]: Ok, I have reproduced this issue locally so far and I will fix this bug in the following days.

Comment (misakicoca) [reply modules/ximgproc/test/test_find_ellipses.cpp:11]: @alalek I have found the source of the error and found a way to fix it, how should I submit the code change?

Comment (alalek) [reply modules/ximgproc/test/test_find_ellipses.cpp:11]: Thank you for looking on this!
Please open a new PR with the fix.

Comment (misakicoca) [reply modules/ximgproc/test/test_find_ellipses.cpp:11]: Thanks very much for your advice! I have submitted the code changes in PR #3393.

**Pull request merged**